cmd/dockerd/daemon.go

```diff
@@ -477,6 +477,17 @@ func (cli *DaemonCli) reloadConfig() {
 		}
 	}
 
+	// On Linux, we use sd_notify to indicate we're reloading config. We send
+	// this signal as early as possible to let systemd know we're reloading,
+	// but must signal "ready" after this completes (even on failure), which
+	// is done by the reload function defined above.
+	//
+	// We could make this more granular (i.e., don't notify if initial
+	// validation failed, and we aborted before actually reloading config.
+	if notifyReloading() {
+		defer func() { notifyReady() }()
+	}
```

Comment:
Using a data dependency on the `notifyReloading` return value to express through the API that the `notifyReady` call needs to be synchronized with it is a brilliant design choice. I suggest going one step further: have `notifyReloading` return a closure which notifies that the reload is done when called. That would have the dual benefit [of making the API both easier to use correctly and harder to misuse](https://blog.codinghorror.com/falling-into-the-pit-of-success/), and of future-proofing the API against other reload-notification protocols we may want to support.
```go
func notifyReloading() (done func())
```
```suggestion
	defer notifyReloading()()
```

Reply:
Ah, I had to run after pushing, and wanted to comment that as alternative, but didn't have the time. Looks like we had the same idea here; I'll update to use that approach 🥹👍